pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1009 @@
+	portBinding := c.portToTCBindings[portName]
+	ofPort := portBinding.ofPort
+
+	portBinding.trafficControls.Delete(tcName)
+	// If the port is no longer used by any TrafficControl, delete the port.
+	if len(portBinding.trafficControls) == 0 {
+		if itf, ok := c.interfaceStore.GetInterfaceByOFPort(ofPort); ok {
+			if err := c.ovsBridgeClient.DeletePort(itf.PortUUID); err != nil {
+				return ofPort, false, err
+			}
+			c.interfaceStore.DeleteInterface(itf)
+		}
+		delete(c.portToTCBindings, portName)
+		return ofPort, true, nil
+	}
+	return ofPort, false, nil
+}
+
+func (c *Controller) syncTrafficControl(tcName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(2).InfoS("Finished syncing TrafficControl", "TrafficControl", tcName, "durationTime", time.Since(startTime))
+	}()
+
+	var tcUpdated bool
+	var returnPortName, targetPortName string
+
+	tc, err := c.trafficControlLister.Get(tcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If the TrafficControl is deleted and the corresponding state doesn't exist, just return.
+			tcState, exists := c.getTrafficControlState(tcName)
+			if !exists {
+				return nil
+			}
+			// If a TrafficControl is deleted but the corresponding state exists, do some cleanup for the deleted
+			// TrafficControl.
+			if err = c.uninstallTrafficControl(tcName, tcState); err != nil {
+				return err
+			}
+			// Delete the state of the deleted TrafficControl.
+			c.deleteTrafficControlState(tcName)
+			return nil
+		}
+		return err
+	}
+
+	// Get the TrafficControl state.
+	tcState, exists := c.getTrafficControlState(tcName)
+	// If the TrafficControl exists and corresponding state doesn't exist, create state for the TrafficControl.
+	if !exists {
+		tcState = c.newTrafficControlState(tcName, tc.Spec.Action, tc.Spec.Direction)
+	}
+
+	if tc.Spec.ReturnPort != nil {
+		// Get name of the return port.
+		returnPortName = c.getPortName(tc.Spec.ReturnPort)
+		// If the name is different from the cached name in the TrafficControl state, it could be caused by the return
+		// port update of the TrafficControl or the creation of the TrafficControl.
+		if returnPortName != tcState.returnPortName {
+			if tcState.returnPortName != "" {
+				// If the stale return port name cached in TrafficControl state is not empty, release the stale return port
+				// from the TrafficControl.
+				returnOFPort, portDeleted, err := c.releaseTrafficControlPort(returnPortName, tcName)
+				if err != nil {
+					return err
+				}
+				// If the stale return port is deleted, uninstall the return flow for the return port.
+				if portDeleted {
+					if err = c.ofClient.UninstallTrafficControlReturnPortFlow(returnOFPort); err != nil {
+						return err
+					}
+				}
+				tcUpdated = true
+			}
+			// Get or create the return port.
+			var returnOFPort uint32
+			if returnOFPort, exists, err = c.getOrCreateTrafficControlPort(tc.Spec.ReturnPort, returnPortName, tcName); err != nil {
```

Comment:
maybe change the second return value to `portCreated` to map `portDeleted` above